File: Data_Structure/DS_Using_Python/Merge_PointofTwo_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		
class findPointOfMerge:
	def __init__(self):
		self.head = None
		
	def printList(self):
		if self.head is None:
			print('List has no element')
			return
		else:
			temp = self.head
			while(temp):
				print(temp.data, end=' ')
				temp = temp.next
		
	def pushData(self, data):
		newNode = Node(data)
		if self.head:
			temp = self.head
			while(temp.next):
				temp = temp.next
			temp.next = newNode
		else:
			self.head = newNode

def countNodes(head):
			cnt = 0
			while head is not None:
				cnt += 1
				head = head.next
			return cnt

def merge(head1, head2):
	cnt1 = countNodes(head1)
	cnt2 = countNodes(head2)
	if cnt1 > cnt2:
		for i in range(0,cnt1-cnt2):
			head1 = head1.next
	else:
		for i in range(0,cnt2-cnt1):
			head2 = head2.next
	while(head1 is not head2):
		head1 = head1.next
		head2 = head2.next
	return head1

#Linked list with even number
LL1 = findPointOfMerge()
LL1.pushData(2)
LL1.pushData(4)
LL1.pushData(6)
LL1.pushData(8)
p = LL1.head.next.next.next

print("Linked List 1 is")
LL1.printList()

#Linked list with odd number
LL2 = findPointOfMerge()
LL2.pushData(1)
LL2.pushData(3)
LL2.pushData(5)
LL2.pushData(7)
q = LL2.head.next.next.next

q.next = p

print("\n\nLinked List 2 is")
LL2.printList()

#Merge function
LL3  = findPointOfMerge()
LL3.head = merge(LL1.head, LL2.head)
print('\n')
if LL3.head is not None:
	print("Merge Point is ",LL3.head.data)
else:
	print("Not Merging")

File: Data_Structure/LinkedList/Merge_PointofTwo_LinkedList.c
//Solved this by Floyd's Cycle-Finding Algorithm or Fast and Slow Pointer method

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
};

struct node *head1 = NULL;
struct node *head2 = NULL;

void linkedListTraversal(struct node* head)
{    
    while(head != NULL)
    {
        printf("%d ", head -> data);
        head = head -> next;
    }
}

void createList(struct node** head, int data)
{
    struct node *ptr = (struct node*)malloc(sizeof(struct node));
    ptr -> data = data;
    ptr -> next = *head;	
    (*head) = ptr;
}

int countNodes(struct node* head)
{
    int cnt = 0;
    while(head != NULL)
    {
        cnt++;
        head = head -> next; 
    }
    return cnt; 
}

struct node* findPointOfMerge4(struct node* h1, struct node* h2)
{ 
    int cnt1 = countNodes(h1);
    int cnt2 = countNodes(h2);
    if(cnt1>cnt2)
    {
        for(int i=0; i<cnt1-cnt2; i++)
        h1 = h1->next; 
    }
    else
    {
        for(int i=0; i<cnt2-cnt1; i++)
        {
            h2 = h2->next;
        } 
    }
    while(h1 != h2)
    {
        h1 = h1->next;
        h2 = h2->next;
    } 
    return h1; 
}

int main()
{
    struct node *p;
    struct node *q;
    
    createList(&head1, 3);
    p = head1;
    createList(&head1, 4);
    createList(&head1, 6);
    createList(&head1, 7);
    createList(&head1, 8);
    createList(&head1, 9);
    printf("Linked List 1 is\n");
    linkedListTraversal(head1);
    
    printf("\n");
    
    createList(&head2, 5);
    q = head2;
    createList(&head2, 7);
    createList(&head2, 8);
    createList(&head2, 2);
    createList(&head2, 1);
    createList(&head2, 10);
    
    printf("\n");
    q -> next = p;
    
    printf("Linked List 2 is\n");
    linkedListTraversal(head2);
    
    printf("\n\nNumber of nodes: %d",countNodes(head1));
    
    struct node* result = findPointOfMerge4(head1, head2);
    if(result != NULL)
    {
        printf("\n\nMerge Point: %d",result -> data); 
    }
    else
    {
        printf("\n\nNot Merging");
    }
    //linkedListTraversal(head);
    return 0;
}